src/main/java/redis/clients/jedis/ScanParams.java

```diff
@@ -1,39 +1,67 @@
 package redis.clients.jedis;
 
+import java.nio.ByteBuffer;
 import static redis.clients.jedis.Protocol.Keyword.COUNT;
 import static redis.clients.jedis.Protocol.Keyword.MATCH;
+import static redis.clients.jedis.Protocol.toByteArray;
+import redis.clients.jedis.Protocol.Keyword;
 
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Collections;
+import java.util.EnumMap;
 import java.util.List;
+import java.util.Map;
 
 import redis.clients.util.SafeEncoder;
 
 public class ScanParams {
-  private List<byte[]> params = new ArrayList<byte[]>();
-  public final static String SCAN_POINTER_START = String.valueOf(0);
-  public final static byte[] SCAN_POINTER_START_BINARY = SafeEncoder.encode(SCAN_POINTER_START);
-
-  public ScanParams match(final byte[] pattern) {
-    params.add(MATCH.raw);
-    params.add(pattern);
-    return this;
-  }
-
-  public ScanParams match(final String pattern) {
-    params.add(MATCH.raw);
-    params.add(SafeEncoder.encode(pattern));
-    return this;
-  }
-
-  public ScanParams count(final int count) {
-    params.add(COUNT.raw);
-    params.add(Protocol.toByteArray(count));
-    return this;
-  }
-
-  public Collection<byte[]> getParams() {
-    return Collections.unmodifiableCollection(params);
-  }
-}
+
+    private final Map<Keyword, ByteBuffer> params = new EnumMap<Keyword, ByteBuffer>(Keyword.class);
+
+    // Default COUNT parameter value according to Redis documentation - http://redis.io/commands/scan
+    private final static ByteBuffer SCAN_COUNT_DEFAULT = ByteBuffer.wrap(toByteArray(10));
+
+    public final static String SCAN_POINTER_START = String.valueOf(0);
+    public final static byte[] SCAN_POINTER_START_BINARY = SafeEncoder.encode(SCAN_POINTER_START);
+
+    public ScanParams match(final byte[] pattern) {
+        params.put(MATCH, ByteBuffer.wrap(pattern));
+        return this;
+    }
+
+    public ScanParams match(final String pattern) {
+        params.put(MATCH, ByteBuffer.wrap(SafeEncoder.encode(pattern)));
+        return this;
+    }
+
+    public ScanParams count(final int count) {
+        params.put(COUNT, ByteBuffer.wrap(Protocol.toByteArray(count)));
+        return this;
+    }
+
+    public Collection<byte[]> getParams() {
+        List<byte[]> paramsList = new ArrayList<byte[]>(params.size());
+        for (Map.Entry<Keyword, ByteBuffer> param : params.entrySet()) {
+            paramsList.add(param.getKey().raw);
+            paramsList.add(param.getValue().array());
+        }
+        return Collections.unmodifiableCollection(paramsList);
+    }
+
+    String match() {
+        if (params.containsKey(MATCH)) {
+            return new String(params.get(MATCH).array());
+        } else {
+            return null ;
+        }
+    }
+    
+    int count() {
```

Comment:
Yes right. I mean it shouldn't have default value. It should not send count parameter instead of sending default value. Default count value for Redis can be changed from Redis side, and we should follow it regardless of Redis versions.